### PR TITLE
Add documentation for memory eviction, update documentation about swap

### PR DESCRIPTION
**Pull request by ahardin-rh**
Continuation of work in  https://github.com/openshift/openshift-docs/pull/2690

Preview build: http://file.rdu.redhat.com/~ahardin/09282016/derekwaynecarr-PR2690/admin_guide/out_of_resource_handling.html

### CODE REVIEW

**Comment by ahardin-rh** [comment]
@adellape please peer review

cc @derekwaynecarr

File: admin_guide/out_of_resource_handling.adoc [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+=== Eviction Monitoring Interval
+
+The node evaluates eviction thresholds per its configured housekeeping interval.
+
+* `housekeeping-interval` is the interval between container housekeepings.
+
+[[out-of-resource-node-conditions]]
+=== Node Conditions
+
+The node will map one or more eviction signals to a corresponding node
```
Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: s/will/can/

File: admin_guide/out_of_resource_handling.adoc [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+The node continues to report node status updates at the frequency specified by
+the `node-status-update-frequency` argument, which defaults to *10s*.
+
+[[out-of-resource-oscillation-of-node-conditions]]
+=== Oscillation of Node Conditions
+
+If a node is oscillating above and below a soft eviction threshold, but not
+exceeding its associated grace period, the corresponding node condition
+constantly oscillates between true and false, which can cause poor scheduling
```
Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: Maybe mark up "true" and "false".

File: admin_guide/out_of_resource_handling.adoc [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+[[out-of-resource-eviction-of-pods]]
+=== Eviction of Pods
+
+If an eviction threshold is met and the grace period is passed, the node
+initiates the process of evicting pods until it observes the signal going below
+its defined threshold.
+
+The node ranks pods for eviction:
+
+. By their quality of service, and
```
Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: Something we can link "quality of service" to explain the available classes? There's a "Quality of Service Classes" in `admin_guide/overcommit`.

Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: Or maybe simply list out the three in a parenthetical or something: BestEffort, Burstable, Guaranteed.

File: admin_guide/out_of_resource_handling.adoc [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+relative to their request are killed first. If no pod has exceeded its request,
+the strategy targets the largest consumer of the starved resource.
+
+A `Guaranteed` pod is guaranteed to never be evicted because of another pod's
+resource consumption. If a system daemon (for example, node, *docker*,
+*journald*, or others) is consuming more resources than were reserved via
+*system-reserved* or *kube-reserved* allocations, and the node only has
+`Guaranteed` pods remaining, then the node must choose to evict a `Guaranteed`
+pod in order to preserve node stability, and to limit the impact of the
+unexpected consumption to other `Guaranteed` pods.
```
Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: If there's a way to break up this long sentence, that'd be nice. Not a super big deal.

File: admin_guide/out_of_resource_handling.adoc [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+[[out-of-resource-eviction-thresholds]]
+=== Eviction Thresholds
+
+The node supports the ability to specify eviction thresholds that trigger the
+node to reclaim resources.
+
+Each threshold is of the following form:
+
+----
+<eviction-signal><operator><quantity>
```
Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: s/eviction-signal/eviction_signal/

**Comment by vikram-redhat** [comment]
Thanks @ahardin-rh . @bfallonf  - Ashley has done the follow up. Can you do a peer review today please?

File: admin_guide/out_of_resource_handling.adoc [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+*journald*, or others) is consuming more resources than were reserved via
+*system-reserved* or *kube-reserved* allocations, and the node only has
+`Guaranteed` pods remaining, then the node must choose to evict a `Guaranteed`
+pod in order to preserve node stability, and to limit the impact of the
+unexpected consumption to other `Guaranteed` pods.
+
+[[out-of-resource-scheduler]]
+=== Scheduler
+
+The node will report a condition when a compute resource is under pressure. The
```
Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: s/will report/reports/

File: admin_guide/out_of_resource_handling.adoc [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+
+|`BestEffort`
+|1000
+
+|`Burstable`
+|min(max(2, 1000 - (1000 * memoryRequestBytes) / machineMemoryCapacityBytes), 999)
+|===
+
+If the node is unable to reclaim memory prior to experiencing a system OOM
+event, the `*oom_killer*` will calculate an `*oom_score*` based on the percentage of
```
Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: s/will calculate/calculates/

**Comment by adellape** [comment]
@ahardin-rh Added minor comments, and we talked about a few things on IRC: adding a link to this from the Release Notes entry on pod eviction, adding an example config snippet early in the topic as to how this stuff is set on the node in general, and adding a link to housekeeping docs.

**Comment by ahardin-rh** [comment]
@adellape Thanks! Edits complete. Please let me know if I missed anything. The Release Notes are updated as well.

File: admin_guide/out_of_resource_handling.adoc [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+compute resource. In cases where this appears to occur, the node can
+proactively fail one or more pods in order to reclaim the starved resource.
+When the node fails a pod, it terminates all containers in the pod, and the
+`*PodPhase*` is transitioned to *Failed*.
+
+Platform administrators can configure eviction settings within the
+*_node-config.yaml_* file:
+
+====
+----
```
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: @derekwaynecarr can you please confirm that this syntax looks okay? Decided to note earlier on in the doc how to configure the settings.

File: admin_guide/out_of_resource_handling.adoc [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+=== Eviction of Pods
+
+If an eviction threshold is met and the grace period is passed, the node
+initiates the process of evicting pods until it observes the signal going below
+its defined threshold.
+
+The node ranks pods for eviction:
+
+. By their xref:../admin_guide/overcommit.adoc#qos-classes[quality of service]
+(`BestEffort`,`Burstable`, or `Guaranteed`), and
```
Comment (adellape) [inline admin_guide/out_of_resource_handling.adoc]: Space before `Burstable`

File: admin_guide/out_of_resource_handling.adoc [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+
+To configure hard eviction thresholds, the following flag is supported:
+
+* `eviction-hard` describes a set of eviction thresholds (for example,
+`memory.available<1Gi`) that, if met, triggers a pod eviction.
+
+[[out-of-resource-eviction-monitoring-interval]]
+=== Eviction Monitoring Interval
+
+The node evaluates eviction thresholds per its configured housekeeping interval.
```
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: @derekwaynecarr how is the housekeeping interval set exactly? Is this another node setting? If so, what does the syntax look like? Thanks!

**Review by adellape — APPROVED** [review]

**Comment by adellape** [comment]
LGTM :+1:

File: admin_guide/out_of_resource_handling.adoc (hunk 9 above)
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: Nice catch! This is updated.

**Comment by bfallonf** [comment]
@ahardin-rh @adellape I think this is fine for the Origin docs, but I'd argue it's not really up to Enterprise standard... I think this can be merged when you guys are ready, and I'll go through it when it's merged. Might be a good chance to have an example of the difference too.

File: admin_guide/out_of_resource_handling.adoc (hunk 8 above)
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: There is an example below in this document.  The argument is either "eviction-hard" or "eviction-soft" so the use of eviction_signal is not accurate.

File: admin_guide/out_of_resource_handling.adoc (hunk 10 above)
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: Hmm, this is an error on my part.  We should just state that it monitors every 10s and the value is not modifiable.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
We need to fix the stanza as desired then this is Lgtm.

**Comment by ahardin-rh** [comment]
@derekwaynecarr @adellape Thanks! This is now updated.

**Comment by adellape** [comment]
:+1:

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../admin_guide/out_of_resource_handling.adoc#admin-guide-out-of-resource-handling[Out of Resource Handling]
|New topic about memory eviction.
%
|xref:../admin_guide/overcommit.adoc#admin-guide-overcommit[Overcommiting]
|Added a Warning box about swap in the xref:../admin_guide/overcommit.adoc#disabling-swap-memory[Disabling Swap Memory] section.
%

**Comment by vikram-redhat** [comment]
@bfallonf / @ahardin-rh - Brice you mentioned that this not suitable in its current form for merging into the enterprise docs, but Ashley has tagged it with enterprise branch for next release, and it is now merged and will make it into the next release. So, I will add a follow up to this till you have had a chance to fix this for the enterprise docs.

**Comment by vikram-redhat** [comment]
Discussed with Brice - he will create a new PR after discussion in a Google doc. I have removed the to_followup flag. So this is now ready to go into the release. If need be, it can be cherry picked this week.

**Comment by vikram-redhat** [comment]
@bfallonf / @ahardin-rh Follow up PR (https://github.com/openshift/openshift-docs/pull/3023) coming soon so this can wait till that is finished. @aheslin

**Pull request merged**